### ISSUE TITLE
Track job attempts individually in PyCondorPlugin

### ISSUE THREAD

**Issue by lucacopa** [dmwm/WMCore#5659]
We find a problem in a tier0 instance: a job was not correctly removed by condor and it remained in status X for a long time. This mess up the accounting for the next retries. 
We are organizing jobs in the condor queue by WMAgentID, but in cases like this, we could have more than one job in condor per WMAgentID.
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/BossAir/Plugins/CondorPlugin.py#L1098
The WMAgent internal job tracking should only ever look at the job attempts that are active. If it kills a job attempt, that attempt has to drop out of tracking. Which means the tracking algorithm should ignore WMAgent job ids and only track glideinwms job ids. 
We should make sure we are protected against this problem in the PyCondorPlugin, I dont think it makes sense to fix the CondorPlugin as it is going to be deprecated soon.

**Comment by hufnagel** [comment]
The BossAir job tracking design is clearly to track by job attempt id. This is returned at submission time and stored. There is no reason to ever look at the WMAgent job id for tracking purposes.

**Comment by tsarangi** [comment]
A check that the job is removed from the queue can be introduced after schedd attempts to remove the job from the queue. It should go after this line:

https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/BossAir/Plugins/PyCondorPlugin.py#L829

For jobs ending in state-X, condor has a force removal mechanism and we can invoke this if the job is still there in the queue after the initial removal is done.

We should think of a way to uniquify the job in the wmagent submit-queue and rely less on remote ids like the glideinwms id that applies only to matched and running jobs...

**Comment by bbockelm** [comment]
Hi,

We should never, never automatically do a "-forcex".  That's a state that needs to be examined by experts.

It should be OK to ignore these jobs after ~30 minutes, however, in terms of making forward progress.

Brian

**Comment by hufnagel** [comment]
I agree and these jobs can be ignored immediately I would think.

Also, trying to get rid of these jobs more vigorous misses the point. If a job drops out of tracking in the agent (because the agent kills it or for whatever reason), it should be out of tracking. What Condor thinks about the job is irrelevant. The agent has internal records what job attempts are active and what are not, these are the authoritative source. If the agent marks a job attempt as failed, that same job attempt should never be considered for anything again, no matter whether it is removed from Condor or stays in state (A, B, C....X) it is in Condor. For all the agent is concerned this is some random job that isn't from that agent.

We don't care about keeping Condor clean, that's a different level of operations and can be followed up on later. We care about getting jobs through and that a previous job attempt sticks around for whatever reason should never stop us from getting the next job attempt through.

WMAgent job != Condor job

**Comment by bbockelm** [comment]
One downside of immediately ignoring these jobs is that they may be still running (in the case of a network disconnect, it takes a bit for both sides to agree to give up).

If it's OK to have two instances of the same job running, then you can immediately ignore.  Otherwise, I'd advise waiting about 30 minutes.

**Comment by hufnagel** [comment]
Then the agent needs to be changed to have a grace period for running job attempts that drop out of tracking. But when the agent marks a job attempt as failed, its failed, there is no going back on that (because it triggers other changes like sending the next job attempt).

**Comment by tsarangi** [comment]
@hufnagel, @lucacopa 
Have a look at the patch ! Let me know if you want to add anything more...

**Comment by hufnagel** [comment]
That might get the job done for the particular problem we had. It doesn't solve the underlying design issue though, it's more like a band aid. A band aid that will be hard to test unless we have a way to trigger a job going into X state (do we ?). Not even sure this fix is worthwhile giving the low probability of jobs going into X state...

It's not really what I wanted though. If you look at

https://github.com/tsarangi/WMCore/blob/master/src/python/WMCore/BossAir/Plugins/LsfPlugin.py#L197

https://github.com/tsarangi/WMCore/blob/master/src/python/WMCore/BossAir/Plugins/LsfPlugin.py#L263

You see I set a gridid parameter that I get from the submission and then use this gridid parameter for the tracking. Neither WMAgent job name nor WMAgent job id is used anywhere in the job tracking. This, by design, avoids any of the problems with multiple attempts of a WMAgent job being active in the system (in any state, not just one active and one X). If we track by gridid, only one of these will be considered by the WMAgent as active. This was always the design, I do not know why it wasn't implemented for the CondorPlugin.

**Comment by hufnagel** [comment]
If I read the current code correctly, it adds the WMAgent job id to the jobs class adds and then does the tracking based on it. Is that correct ? If so, that is not unique, condor doesn't care about this being a unique parameter. Which is ultimately where our problem here comes from.

If we followed the BossAir design, we would have to get the condor id of a submitted job right after submission, update the BossAir job record with it (the code for that exists already as you can see in the LSFPlugin) and then track based on the condor id.

I realize this could be quite a big change. Two questions to determine what type of change this would be:

1) Can we easily get the condor id of a newly submitted job ? Ideally you want feedback from the submission itself, looking it up later through some other non-unique id (wmagent job name/id) doesn't really help as you just move the point of failure.

2) Can the job tracking code be adapted to use condor id, ie. is that information already available in the information we extract from condor for tracked jobs ?

**Comment by tsarangi** [comment]
@hufnagel 
We can get the condor job id once the job is submitted to the queue through the same mechanism as the getClassAds.  The real problem lies in how to map it to the WMA job without going through the WMA parameter like the WMAgentID that is getting fed to the DB during the job creation. 

I am not that old in this business to know why we chose the CondorPlugin code to be like this while the BossAir design was aimed to be something different !

I understand the improvement you are talking about, but what I don't understand is why now ? In all these years of running with CondorPlugin, if a problem related to this popped out only once, then why change the design ! I would like to see the same problem occurring several times more before changing it.

For now, bandaid will heal the not-so-deep wound :-)

**Comment by hufnagel** [comment]
Why change now ? Why wait if there is a problem ? The BossAir design is cleaner and protects not only against this issue but a multitude of other issues, all related to the simple fact that there is no guarantee that there is a one to one mapping between WMAgent jobs and Condor jobs. There isn't even a guarantee that there is a one to one mapping between WMAgent jobs and active (like normal, not abandoned, held, removed) Condor jobs. Once you go into high failure modes and pre-emption (which will happen a lot on the HLT for instance), a mix of frequent internal Condor retries and WMAgent retries will stress this part of the system.

I have experience with this from the Run1 Tier0 operations. The LsfPlugin back then relied on job tracking by job name (which included the job id). This used to work 99% of the time but had all these little edge cases where the one to one mapping between Tier0 job and LSF job broke down. That then always caused weird symptoms because the system wasn't designed to deal with it. It never really paid to debug these, because the problems were basically not reproducible. But they did cause operational load. So I rewrote the LsfPlugin to avoid these problems from the get-go, which in the end was a minimal amount of work and well worth the operational savings.

Hm, so the Condor python submit API call does not return the Condor job id then ? Even without that, we don't need the internal Condor id to be able to track a Condor job. We can just assign our own, as long as it is unique. Could be as simple as

gridid = '%i_%i' % (jobID, uuid.uuid1())

and then in the job adds replace

WMAgent_JobID == '%i' % jobID

with

WMAgent_GridlD == '%s' % gridid

and also adjust the tracking code accordingly.

We don't use the WMAgent job id in the jdl for anything but tracking, right ?

Also, we have not switched to the PyCondorPlugin in production. There needs to some validation campaign before that can happen. We want to make any large changes now so we only have to do that validation once.

**Comment by bbockelm** [comment]
Hi,

The condor python submit API returns the cluster ID; the process ID is 0 by default (and 0...N-1 for the N jobs submitted).

However, I see no two-phase commit protocol here.  Dirk's suggestion appears to likely result in multiple running HTCondor jobs per WMAgent job when a crash occurs after a HTCondor submit but before the HTCondor job ID is committed to disk.

So, it appears neither option is safe (nor the existing CondorPlugin).  Both can result in the situation Dirk is worried about.

Let me suggest one of two consistency protocols:
1) Make the job idempotent - i.e., it's safe to have multiple copies run.
2) Use the schedd.transaction object to do a two phase commit - only commit the HTCondor job after HTCondor job ID has been committed to the WMAgent's DB.

(1) is going to have better scaling characteristics.  (2) is going to be safer in more cases.

Brian

**Comment by tsarangi** [comment]
Before we go ahead and promise ourselves to make big changes to the code, I would like to understand what are the other possible cases where 2 or more instances of the same job will coexist at the same time in the submit queue and can't be handled by the 6 or so JOBStatus codes provided by the schedd.

Furthermore, i think this particular issue 5659 is solved by the patch provided and a new issue should be opened to address the improvement under discussion.

This should not hold back the PyCondorPlugin to go into production because it is already better than the existing CondorPlugin in multiple ways.

**Comment by hufnagel** [comment]
@bbockelm , the situation I worry about is multiple job attempts for the same WMAgent job in the system *and* that situation confusing WMAgent. (1) by itself will fix that as it solves the latter problem.

Without (2) you can still end up with multiple copies of a job, *but* they should not lead to any problems on the WMAgent side and therefor should only be a resource utilization efficiency problem. Which shouldn't be a big issue as I expect the scale of the problem to be small (if the submission regularly crashes between submit and recording that submit in the WMAgent, we have larger fish to fry then some rogue runaway jobs).

@tsarangi , I would agree with your suggestion, except for the fact that the solution here is untested and basically almost untestable since we cannot easily trigger the conditions that caused the problem in the first place. Anyone have an idea how we can test this ? Best we can probably do is to test that it doesn't interfere with normal operations and then merge it, hoping that it also fixed the problem case.

**Comment by tsarangi** [comment]
We can test this by killing the condor shadow process of the job on the submit machine that should send it to X state.
